server/services/user.service.js

```diff
@@ -32,18 +32,56 @@ export const comparePassword = async (password, hashedPassword) => {
   }
 };
 
+export const getUserByReferralCode = async (referralCode, fields = null) =>
+  User.findOne({ referralCode }).select(fields);
+
+export const generateReferralCode = async (firstName) => {
+  const name = firstName.substring(0, 3).toLowerCase();
+  let code = name;
+  let charsToAdd = 4;
+
+  if (firstName.length < 3) {
+    charsToAdd = 7 - firstName.length;
+  }
+
+  function generateCode(pre) {
+    const possible = '0123456789';
+    let fullCode = pre;
+    for (let i = 0; i < charsToAdd; i += 1) {
+      fullCode += possible.charAt(Math.floor(Math.random() * possible.length));
+    }
+    return fullCode;
+  }
+
+  code = generateCode(code);
```

Comment:
This would be better as something like this depending on how you resolved the previous comments I left.
```suggestion
  const referralCode = generateCode(name);
```